### ISSUE TITLE
Wrong wrapping of contour points using dense deformation

### ISSUE THREAD

**Issue by KaiXin21** [voxelmorph/voxelmorph#384]
Hello, 
**Task**
I am trying to use VoxelMorph (VM) to register cardiac ultrasound images.

**What have you tried**
I trained a baseline VM model without diffeomorphic or probabilistic, and predict the displacement field between end-systolic (ES) frame and end-diastolic (ED) frame. 

**Details of experiments**
I get moved_image by wrapping the moving_image using the resulted dense displacement field. I also want to wrap the contour (annotation points). The Red curve is the contour of left ventricle (LV) of fixed_image, the blue curve is obtained by wrapping the red curve.  I was thinking: the wrapping of contour in the fixed_image would match the moving_image, but it seems not doing so. 
From ES to ED, comparing the moving and moved images, I can see the LV is obviously enlarged, however, the warped contour (blue curve) does not obviously shrinked… did I wrap the contour correctly?

![image](https://user-images.githubusercontent.com/96538810/148211043-bb0aa209-8657-4b7c-8c55-835062162a49.png)

I use these code:
![image](https://user-images.githubusercontent.com/96538810/148211201-9169275e-933e-4cb4-b967-2f1baf99e6b0.png)

Thanks a lot for helping out!

**Comment by adalca** [comment]
just to make sure, you mean [warping](https://en.wikipedia.org/wiki/Image_warping) not [wrapping](https://en.wikipedia.org/wiki/Wrapping_(graphics)), right?

I can't parse all the code, but what I can see looks okay to me. To test this code with a simpler example, I would recommend trying a very quick MNIST model and then warping some points in those images with this exact code you showed here.

It's quite possible that the warped image "looks" good but it's actually a bad registration (this is why we look at annotations!). Another way to understand if this might be happening is to look at the deformation field itself. I would do this by warping an image grid (pystrum.pynd.ndutils.bw_grid) with the deformation field and seeing if it's smooth and if it makes sense given your images.

**Comment by KaiXin21** [comment]
@adalca 
Thank you for helping, I try a test on the MNIST data, but still have annotation warping errors. Yes, I mean warping not wrapping. 
I trained a model on the MNIST data following the tutorial on https://colab.research.google.com/drive/1WiqyF7dCdnNBIANEY80Pxw_mVz4fyV-S?usp=sharing#scrollTo=joVczQLTPXMZ. 
Train 10 epochs, the loss looks decaying:
![image](https://user-images.githubusercontent.com/96538810/148416898-ff13dd71-0405-4f7b-8010-68cbd06175b6.png)
I use the trained model to predict deformation between a moving and fixed image pair, and deformed the moving_image to get the moved_image, which looks fine:
![image](https://user-images.githubusercontent.com/96538810/148416970-5d624f6a-810e-45a7-bc34-98f738259f5e.png)
I created some key points (Red points) on the fixed_image,  the blue points are obtained by warping the red points using the exact code I post here. It seems the blue points not quite matching the moving_image. Could it be the problem of bad registration model or the annotation warping  code(method)?
![image](https://user-images.githubusercontent.com/96538810/148417045-5949f6d0-99ad-4aea-8ad1-9936f47fcae0.png)
The code I use:
![image](https://user-images.githubusercontent.com/96538810/148417138-b02e896d-bb07-4a39-9d7c-c10fbf429285.png)
Another problem I found is: when I run "vxm.utils.point_spatial_transformer(data)[0, ...].numpy()" on Windows PC, it got error:

  File "C:\Users\Anaconda3\envs\voxelMorph\lib\site-packages\tensorflow\python\framework\constant_op.py", line 96, in convert_to_eager_tensor
    return ops.EagerTensor(value, ctx.device_name, dtype)

ValueError: Failed to convert a NumPy array to a Tensor (Unsupported numpy type: NPY_INT).

When I run the same code on a Linux PC using same conda environment as windows, it's fine. How can I make it works on windows?

I am still working on how to do warping image grid…before this,  I try to visualize the flow, just found it hard to interpret as not knowing the location correspondence  between the myocardium and the flow vectors….. By setting axis on, I can see the origin of image is at upper left corner, while the origin of flow is at lower left corner,  are they presenting the same view/orientation? 
![image](https://user-images.githubusercontent.com/96538810/148417573-2e110926-0ac8-42f0-90d7-5ec2cb6bba6b.png)
Thank you very much, sorry this maybe a bit more…

**Comment by adalca** [comment]
ah, at a quick glance, one small issue might be that plotting work in [col, row] whereas numpy/tf/etc work with [row, col] indexing. 
you can read about ij vs xy indexing here: https://numpy.org/doc/stable/reference/generated/numpy.meshgrid.html

But basically what you can do is flip the [x, y] coordinated in your `annotation_fix`
Then when you plot show annotation_fix[:, 1] first
same with annotation_mov

I think this should work. Give it a try with MNIST and lmk how it looks?

**Comment by KaiXin21** [comment]
@adalca 
Thanks! I tried flipping the [x, y] coordinated for MNIST, it works well!!
![image](https://user-images.githubusercontent.com/96538810/148567653-483d064e-4fc4-47f8-a6fa-aa345bdbd2fe.png)
And I try the same trick on the heart image,  now the warped contour (blue curve) have obviously shrinked!
![image](https://user-images.githubusercontent.com/96538810/148568665-c7afeaef-098c-4434-9a1d-99b4b41904a9.png)

Sorry can I ask one more question here, or shall I open a new issue?  If warp an image several times using several displacement fields, will the resulted warped image very blurred?  After 10 times of warping I got this:
![image](https://user-images.githubusercontent.com/96538810/148568783-a8a2604d-59d9-496d-adb0-72bf8f671235.png)

If I have several displacement fields, should I first compose all these displacement fields then warp the image with the composed transformation? Does VoxelMorph has functions to compose several displacement fields?

Thanks a lot for your kind help!!

**Comment by adalca** [comment]
Yup, if you warp several times with the default (linear) interpolation, you will keep blurring the image a bit more every time. You should definitely compose the transformations (we have utilities for this in `vxm.utils` and layers for composing as well): https://github.com/voxelmorph/voxelmorph/blob/f14939a6db68e14b0e1b1fefe4b97ce90852809e/voxelmorph/tf/utils/utils.py#L215

**Comment by KaiXin21** [comment]
Many thanks again for the suggestions, they are very helpful guidance!! let me try to compose the transforms.

**Comment by KaiXin21** [comment]
@adalca 
Can I ask where is the function of "affine_to_shift" in the latest Github version of voxelMorph?  I've installed latest voxelmorph from Github, and found 
AttributeError: module 'voxelmorph.tf.utils' has no attribute 'affine_to_shift'...
Thank you!

**Comment by adalca** [comment]
Where did you get this error?

the function is here, and slightly different named: https://github.com/voxelmorph/voxelmorph/blob/67bc9d209c2eb42dee09276b5d1cb77eb8587c54/voxelmorph/tf/utils/utils.py#L565

@ahoopes might also be of help

**Comment by KaiXin21** [comment]
@adalca @ahoopes 
I have installed the latest Github version of VoxelMorph, and trying to do annotation warp.
The AttributeError: module 'voxelmorph.tf.utils' has no attribute 'affine_to_shift'   pop up when I try to run 
"field_inv = vxm.utils.affine_to_shift(aff_inv, im.shape, shift_center=True)[np.newaxis, 
...]" in VoxelMorphAffineAnnotationWarp.ipynb (https://colab.research.google.com/drive/1V0CutSIfmtgDJg1XIkEnGteJuw0u7qT-#scrollTo=fXIu9Na4aEeW). 

If I replace 'affine_to_shift' with 'affine_to_dense_shift',  the following pop up: 

 File "C:\Users\Anaconda3\envs\VM\lib\site-packages\voxelmorph\tf\utils\utils.py", line 500, in validate_affine_shape
    raise ValueError(f'Affine matrix must be of shape (2, 3) or (3, 4), got {actual}.')

ValueError: Affine matrix must be of shape (2, 3) or (3, 4), got (3, 3).

Many thanks for helps.

**Comment by ahoopes** [comment]
In vxm utilities, the affine matrix shape should always be `(N, N+1)`, where `N = ndim`, for consistency. So you can crop your 2D input affine matrix to shape (2, 3) with `matrix[:2, :]`. That should fix the error.

**Comment by KaiXin21** [comment]
@adalca, sorry to bother you again, I try to visualize the deformation field by warping an image grid and compute the determinant of Jacobian. Here is the result:
First row shows: deformed moving_image and grid using deformation_field1 (∅1), and determinant of Jacobian of ∅1.
Second row shows: deformed moving_image and grid using deformation_field2 (∅2), and determinant of Jacobian of ∅2.  ∅2 is a composed of several displacement fields coming from inter-frame registrations.  
![image](https://user-images.githubusercontent.com/96538810/152525673-60c358cc-d44d-4a01-886b-e3b9e8fd0c67.png)
I found two observations are a bit counter-intuitive:
1.   ∅2 seems less smooth and more irregularized than ∅1, however, ∅2  has much lower percentage of folding pixels, is it possible or might be error? Which one is better registration, ∅1 or ∅2? 
2.   Looking at the region pointed with blue arrow in the grid figure, the grids are pushing into the center, I though it is compressing; however, this region has     determinant of Jacobian |𝐽|  greater than 1,  |𝐽| >1 indicate local expansion… ? Why this region seems compressing but with |𝐽| shows expanding? 

Maybe I miss understood the ∅ and |J|,  Thank you for your time.
